#Online Payments Fraud Detection









Connecting Google Colab to Kaggle to get Dataset directly to colab

In [ ]:
!kaggle datasets download -d jainilcoder/online-payment-fraud-detection

Downloading the helper functions designed by mrdbourke which contains custom functions

In [ ]:
!wget https://raw.githubusercontent.com/mrdbourke/tensorflow-deep-learning/main/extras/helper_functions.py

Installing pycaret library

In [ ]:
!pip install --pre pycaret

Installing model deployment library

In [ ]:
!pip install gradio

In [ ]:
!pip install -U dataprep

Importing required functions from helper functions

In [8]:
from helper_functions import unzip_data

unziping the dataset

In [9]:
unzip_data('online-payment-fraud-detection.zip')

Importing required libraries

In [10]:
import pandas as pd
import numpy as np
from pycaret.classification import *
from dataprep.eda import create_report
import matplotlib.pyplot as plt
%matplotlib inline

In [11]:
df = pd.read_csv("onlinefraud.csv")

In [12]:
df.head()

step      type    amount     nameOrig  oldbalanceOrg  newbalanceOrig  \
0     1   PAYMENT   9839.64  C1231006815       170136.0       160296.36   
1     1   PAYMENT   1864.28  C1666544295        21249.0        19384.72   
2     1  TRANSFER    181.00  C1305486145          181.0            0.00   
3     1  CASH_OUT    181.00   C840083671          181.0            0.00   
4     1   PAYMENT  11668.14  C2048537720        41554.0        29885.86   

      nameDest  oldbalanceDest  newbalanceDest  isFraud  isFlaggedFraud  
0  M1979787155             0.0             0.0        0               0  
1  M2044282225             0.0             0.0        0               0  
2   C553264065             0.0             0.0        1               0  
3    C38997010         21182.0             0.0        1               0  
4  M1230701703             0.0             0.0        0               0

checking datatypes and size of dataset

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6362620 entries, 0 to 6362619
Data columns (total 11 columns):
 #   Column          Dtype  
---  ------          -----  
 0   step            int64  
 1   type            object 
 2   amount          float64
 3   nameOrig        object 
 4   oldbalanceOrg   float64
 5   newbalanceOrig  float64
 6   nameDest        object 
 7   oldbalanceDest  float64
 8   newbalanceDest  float64
 9   isFraud         int64  
 10  isFlaggedFraud  int64  
dtypes: float64(5), int64(3), object(3)
memory usage: 534.0+ MB


Sampling the data

* As the dataset too large which consists around 6362620 entries, if we procced entire dataset at one go ,it will take lot of processing time.

* To solve this issue we going to take sample out of this dataset , i have taken 20% of data as sample, you can tune this sample size



In [14]:
df1 = df.sample(frac=0.2)

Dropping unnecessary columns 

In [15]:
df1 = df1.drop(columns = ['nameOrig','nameDest','isFlaggedFraud'])

Over Sampling Method

* As the Target Variable i.e isFraud contains less no of companies goes isFraud i.e no of 1 is less as compared to 0. This is where we use oversampling method to make this two classes balanced.

In [16]:
df1['isFraud'].value_counts()

0    1270913
1       1611
Name: isFraud, dtype: int64

#Model Building using Auto ML



> Pycaret AutoML Library

* PyCaret is an open-source, low-code machine learning library in Python that automates machine learning workflows.

* It is an end-to-end machine learning and model management tool that exponentially speeds up the experiment cycle and makes you more productive.

* Compared with the other open-source machine learning libraries, PyCaret is an alternate low-code library that can be used to replace hundreds of lines of code with a few lines only. This makes experiments exponentially fast and efficient.



Step 1 : Setting up preprocessing the data: 

* fix_balance : over sampling
* remove_outliers: outliers
* target : dependent variable

In [17]:
clf1 = setup(data=df1, target='isFraud',fix_imbalance=True,remove_outliers=True,use_gpu=True)

#Exploratory Data Analysis 

In [29]:
rep = create_report(df1)
rep.save("EDA.html")

Report has been saved to EDA.html!


check which model perform best

In [19]:
best_model = compare_models()

Processing:   0%|          | 0/61 [00:00<?, ?it/s]

Logistic Regression has performed really well, so im going to use this model for deployment

In [21]:
et = create_model('et')

Processing:   0%|          | 0/4 [00:00<?, ?it/s]

In [22]:
predict_model(et)

step  type_CASH_OUT  type_PAYMENT  type_CASH_IN  type_TRANSFER  \
890766   259.0            1.0           0.0           0.0            0.0   
890767    41.0            0.0           0.0           1.0            0.0   
890768   236.0            1.0           0.0           0.0            0.0   
890769   428.0            0.0           1.0           0.0            0.0   
890770   215.0            0.0           0.0           0.0            1.0   
...        ...            ...           ...           ...            ...   
1272519  573.0            0.0           1.0           0.0            0.0   
1272520   40.0            1.0           0.0           0.0            0.0   
1272521  422.0            0.0           1.0           0.0            0.0   
1272522  356.0            1.0           0.0           0.0            0.0   
1272523  303.0            0.0           1.0           0.0            0.0   

         type_DEBIT         amount  oldbalanceOrg  newbalanceOrig  \
890766          0.0  124542.101562       0.000000         0.00000   
890767          0.0   83887.093750  605155.812500    689042.87500   
890768          0.0  115140.609375    9922.000000         0.00000   
890769          0.0    9637.040039       0.000000         0.00000   
890770          0.0  702024.750000   50122.000000         0.00000   
...             ...            ...            ...             ...   
1272519         0.0    3660.110107   89068.578125     85408.46875   
1272520         0.0  107675.882812   10811.000000         0.00000   
1272521         0.0   43653.039062       0.000000         0.00000   
1272522         0.0  363220.437500       0.000000         0.00000   
1272523         0.0    5190.890137       0.000000         0.00000   

         oldbalanceDest  newbalanceDest  isFraud  prediction_label  \
890766     1.464258e+06    1.656980e+06        0                 0   
890767     5.269539e+05    4.430668e+05        0                 0   
890768     2.817763e+05    3.969169e+05        0                 0   
890769     0.000000e+00    0.000000e+00        0                 0   
890770     6.699871e+05    1.372012e+06        0                 0   
...                 ...             ...      ...               ...   
1272519    0.000000e+00    0.000000e+00        0                 0   
1272520    5.780000e+02    1.060969e+05        0                 0   
1272521    0.000000e+00    0.000000e+00        0                 0   
1272522    4.972458e+05    8.604662e+05        0                 0   
1272523    0.000000e+00    0.000000e+00        0                 0   

         prediction_score  
890766               1.00  
890767               1.00  
890768               1.00  
890769               1.00  
890770               1.00  
...                   ...  
1272519              1.00  
1272520              0.95  
1272521              1.00  
1272522              1.00  
1272523              1.00  

[381758 rows x 14 columns]

In [23]:
save_model(et,'best model fraud detection')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=/tmp/joblib),
          steps=[('numerical_imputer',
                  TransformerWrapper(exclude=None,
                                     include=['step', 'amount', 'oldbalanceOrg',
                                              'newbalanceOrig', 'oldbalanceDest',
                                              'newbalanceDest'],
                                     transformer=SimpleImputer(add_indicator=False,
                                                               copy=True,
                                                               fill_value=None,
                                                               missing_values=nan,
                                                               strategy='mean',
                                                               verbose=0))),
                 ('categorical_imputer',
                  Transform...
                  ExtraTreesClassifier(bootstrap=False, ccp_alpha=0.0,
              

In [24]:
loaded_model = load_model("best model fraud detection")

Transformation Pipeline and Model Successfully Loaded


Running app for actual prediction

In [26]:
create_app(loaded_model)

Colab notebook detected. To show errors in colab notebook, set debug=True in launch()
Note: opening Chrome Inspector may crash demo inside Colab notebooks.

To create a public link, set `share=True` in `launch()`.


<IPython.core.display.Javascript object>